### PR TITLE
Support for Short Codes and Alphameric sender ID

### PR DESCRIPTION
**Pull request by Marahin**
Tested on example app:

```
Sent text to +48 12 545 67 89
Texter: User#welcome
Date: 2017-11-24 15:39:54 +0100
From: Team Sp.z.o.o
To: +48 12 545 67 89
Rendered user_texter/welcome.text.erb (0.2ms)
```

Keep in mind that in order to use Alphanumeric sender ID you have to declare it this way: `from: 'My Team <Team Inc.>'`, where 'My Team' will be equal to name, and `Team Inc.` to Sender ID (`.from_phone`).

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,3 +1,5 @@
 source 'https://rubygems.org'
 
 gemspec
+
+gem 'pry'
```
Comment (Lackoftactics) [inline Gemfile]: I am not sure if pry should be here

File: lib/textris/message.rb [hunk 2]
```diff
@@ -81,15 +81,20 @@ def parse_from(from)
     end
 
     def parse_from_dual(from)
-      if (matches = from.to_s.match(/(.*)\<(.*)\>\s*$/).to_a).size == 3 &&
-          Phony.plausible?(matches[2])
+      matches = from.to_s.match(/(.*)\<(.*)\>\s*$/).to_a
+      return if matches.size != 3
```
Comment (Lackoftactics) [inline lib/textris/message.rb]: magic number, what 3 represents here?

Comment (Marahin) [reply lib/textris/message.rb]: @Lackoftactics https://github.com/visualitypl/textris/blame/ab770cb966218d6a4c2225a7a7ab360d590698ae/lib/textris/message.rb#L88

But:

```
irb(main):001:0> "Jan Matusz <+48123456789>".match(/(.*)\<(.*)\>\s*$/).to_a
=> ["Jan Matusz <+48123456789>", "Jan Matusz ", "+48123456789"]
```

So basicly it "splits" the sender string (which allows us to normalize/set the sender ID and sender name explicitly).

File: lib/textris/message.rb [hunk 3]
```diff
@@ -81,15 +81,20 @@ def parse_from(from)
     end
 
     def parse_from_dual(from)
-      if (matches = from.to_s.match(/(.*)\<(.*)\>\s*$/).to_a).size == 3 &&
-          Phony.plausible?(matches[2])
+      matches = from.to_s.match(/(.*)\<(.*)\>\s*$/).to_a
```
[the remark this answers is not in the document]
Comment (Marahin) [reply lib/textris/message.rb]: @Lackoftactics done, but this required me to use the short try notation (`&.`).

File: lib/textris/message.rb [hunk 4]
```diff
@@ -81,15 +81,20 @@ def parse_from(from)
...
+      return if matches.size != 3
+      if Phony.plausible?(matches[2]) || PhoneFormatter.is_a_short_code?(matches[2])
```
Comment (Lackoftactics) [inline lib/textris/message.rb]: maybe some simple helper method, that will be readable for getting name and number from the dual field

File: lib/textris/message.rb [hunk 5]
```diff
@@ -81,15 +89,23 @@ def parse_from(from)
     end
 
     def parse_from_dual(from)
-      if (matches = from.to_s.match(/(.*)\<(.*)\>\s*$/).to_a).size == 3 &&
-          Phony.plausible?(matches[2])
-        [matches[1].strip, Phony.normalize(matches[2])]
+      matches = from.match(/(.*)\<(.*)\>\s*$/)
+      return if matches.nil?
```
Comment (Lackoftactics) [inline lib/textris/message.rb]: checking nil? maybe better to

`return if !matches`

What do you think?

Comment (Marahin) [reply lib/textris/message.rb]: @Lackoftactics maybe `return unless matches`?

File: lib/textris/phone_formatter.rb [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+      def is_a_phone_number?(phone)
+        Phony.plausible?(phone)
+      end
+
+      # We ASSUME that if someone passes a value
+      # that cannot be resolved as a short code or a phone
+      # number, then it is Alphameric Sender ID
+      def is_alphameric?(phone)
+        !is_a_phone_number?(phone) && !is_a_short_code?(phone)
```
Comment (philnash) [inline lib/textris/phone_formatter.rb]: [The rules for an alphanumeric ID](https://support.twilio.com/hc/en-us/articles/223181348-Getting-started-with-Alphanumeric-Sender-ID) are:

> You may use any combination of 1 to 11 letters, A-Z and numbers, 0-9. Both lowercase and uppercase characters are supported as well as spaces. 1 letter and no more than 11 alphanumeric characters may be used.

I messed around with a regex to match this constraint and came up with:

```
\A(?=.*[a-zA-Z])[a-zA-z\d]{1,11}\z
```

Broken down, this is:

```
\A               # Start of the string
(?=.*[a-zA-Z])   # Lookahead to ensure there is at least one letter in the entire string
[a-zA-z\d]{1,11} # Between 1 and 11 characters in the string
\z               # End of the string
```

So we could rewrite this as:

```ruby
def is_alphanumeric?(phone)
  !!phone.to_s.match(/\A(?=.*[a-zA-Z])[a-zA-z\d]{1,11}\z/)
end
```

File: spec/textris/phone_formatter_spec.rb [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  it 'should recognise a 5 digit short code' do
+    expect(described_class.format("44397")).to eq('44397')
+  end
+
+  it 'should recognise a 6 digit short code' do
+    expect(described_class.format("443975")).to eq('443975')
+  end
+
+  it 'treat non-short code non-phone number numbers as alphamerics' do
+    ['8945', '894', '89', '8', '8945467', '89454678', '894546789'].each do |number|
```
Comment (philnash) [inline spec/textris/phone_formatter_spec.rb]: These tests are incorrect, alphanumeric sender IDs are required to have at least one letter.

Comment (Marahin) [reply spec/textris/phone_formatter_spec.rb]: This test case was not supposed to check whether it's classified as alphameric, but rather 'treated' as it was like it (so no additional formatting), if I remember correctly :) but yes, this should be replaced with a proper alphamerics tests, I agree.

File: lib/textris/phone_formatter.rb (hunk 6 above)
Comment (Marahin) [reply lib/textris/phone_formatter.rb]: Awesome!

**Pull request merged**